### ISSUE TITLE
How could I load image from local Source Storage?

### ISSUE THREAD

**Issue by li3cmz** [heartexlabs/label-studio#774]
Hi, I have followed the doc [here](https://labelstud.io/guide/tasks.html#Import-data-from-a-local-directory) to create a local storage.
![a](https://user-images.githubusercontent.com/32973240/114020353-1470e900-98a2-11eb-8ceb-41605806ea99.png)

And then I start to use the local storage by two steps.

1. Put .*png file into the local path that I select above.
2. Import my tasks through the json file below with the image value("original_image") inside the red box.
![b](https://user-images.githubusercontent.com/32973240/114020362-18047000-98a2-11eb-811d-d636f4df2088.png)

However, I failed to load the image through this way as below.
![c](https://user-images.githubusercontent.com/32973240/114020373-1a66ca00-98a2-11eb-915a-d3db15fa9eab.png)

Could you help me with the bug? Looking forward to your reply.

**Comment by makseq** [comment]
Hi! 
For Local files storage you have to use `/data/local-files/1.png`.

**Comment by makseq** [comment]
@smoreface Could you add to this doc section 
https://labelstud.io/guide/tasks.html#Import-data-from-a-local-directory
a task example with the correct path to file (and maybe some note about `/data/local-files/` path), like: 

```
{
 "data": {
    "image": "/data/local-files/1.jpg"
  }
}

**Comment by li3cmz** [comment]
> Hi!
> For Local files storage you have to use `/data/local-files/1.png`.

Hi, I tried your suggestion but failed again as below. 
![e](https://user-images.githubusercontent.com/32973240/114121365-d36be980-9920-11eb-9cce-f66b409b1ce0.png)

I want to know the requirement of setting the local path while adding the source storage. It is the absolute path of starting from the home directory as below.
![d](https://user-images.githubusercontent.com/32973240/114121303-b505ee00-9920-11eb-86db-bbbc3ccc8160.png)

**Comment by Morgadoooo** [comment]
Hi, 

Isn't it suppose to be /data/local-files/?d=1.png 
Or am i misunderstanding something 

@makseq @li3cmz

**Comment by makseq** [comment]
@Morgadoooo yes, of course. You are right!

**Comment by Morgadoooo** [comment]
> > Hi!
> > For Local files storage you have to use `/data/local-files/1.png`.
> 
> Hi, I tried your suggestion but failed again as below.
> ![e](https://user-images.githubusercontent.com/32973240/114121365-d36be980-9920-11eb-9cce-f66b409b1ce0.png)
> 
> I want to know the requirement of setting the local path while adding the source storage. It is the absolute path of starting from the home directory as below.
> ![d](https://user-images.githubusercontent.com/32973240/114121303-b505ee00-9920-11eb-86db-bbbc3ccc8160.png)

Hi, if i'm not mistaking i got this info from : core.settings.base

# Build paths inside the project like this: os.path.join(BASE_DIR, ...)
BASE_DIR = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))

So i think you can probably specify path regarding this info of BASE_DIR

**Comment by li3cmz** [comment]
Hi, if i'm not mistaking, I think that there maybe some inappropriate and misleading codes for [get_data](https://github.com/heartexlabs/label-studio/blob/fff1de86bfe644759f30e0c16cc211039c608ef9/label_studio/io_storages/localfiles/models.py#L54). 

Maybe the environment variable `LOCAL_FILES_DOCUMENT_ROOT` should be replaced by `self.path` which is received from the Label Studio UI. And then we don't have to set the `LOCAL_FILES_DOCUMENT_ROOT` again. 

At first I thought that there was no need to set `LOCAL_FILES_DOCUMENT_ROOT` again after setting `Local path` in the UI of **Edit Source Storage**, which caused the error that I mentioned above.

And the error appeared after I set the variable `use_blob_urls` to be true as the screenshot shown below.
![WXWorkCapture_16185623512878](https://user-images.githubusercontent.com/32973240/114999516-82d22e80-9ed4-11eb-8e67-fc8a9affadb2.png)

Maybe the problem could be solved by more specific documentation for `LOCAL_FILES_DOCUMENT_ROOT` or changing the code.

**Comment by zakajd** [comment]
@li3cmz Did you manage to import local files without errors?
What did you change in the initial example?

**Comment by savhascelik** [comment]
@dsantiago absolutely I agree but  I found a solution like this. First of all, after setting the local source, selecting the "Treat every bucket object as a source file" option and including the files -"Sync Storage"- I learn the path of the files and change the paths of the json files I include.

**Comment by AlcobaAI** [comment]
I just had this same issue and fixed it by formatting the path to each image like this: data/local-files/?d=home/path/to/image.jpg .

**Comment by yazansayed** [comment]
I just want to express disappointment for handling such a necessary task as loading data from a local json, wasted a lot of time because of this

**Comment by Reasat** [comment]
The above answers seems to have a lot of confusion. Is there a simple answer to the question, "How to add images from a local directory?"

**Comment by frippe75** [comment]
Ahh.. Just got back to trying to eval label studio after leaving it the last time since it consumed ALL my time set a side for a project rather than moving the project forward focusing on the "value-add". This should be the main focus on any type of tool. Back again with some fresh energy. Stuck again trying to run locally using docker and local files. This time I'm restricting my eval time to 10 hours, soon to be out-of-them...

I restorted to using custom jupyter notebook and winging my own. Dataset quickly grew to 50.000 labeled images.

Now I'm stuck with 15 COCO dataset files. Used the label-studio-converter and tried multiple 
--image-root-url way of trying to get them to match but getting 403 and that I cannot even read the /data/local-files directory... There is an old issue where people use a chown -R during startup...

I think there should be a dead-simple way of getting up and running. Tried a conda env and pip install ... Never completes successfully and takes for ever. Label Studio looks like a great option and I would really like to get started using it.

**Comment by unfa** [comment]
Let me be another one to say: I am very excited about Label Studio and creating my first dataset and CV model, but I spent hours trying to add my locally stored images and it seems like everyone says something different, and the docs make it look rather easy, but I get errors constantly and can't progress though this. Do I really need to put my local data into an S3 bucket to get going?

1. I installed via pip in a dedicated Python `venv`;
2. I exported the designated paths to environment before running `label-studio`;
3. I run the helper web server as documented, but it never adds anything to the `files.txt` file. It's empty. Trying to specify a wildcard also produces errors from `find` command about missing or wrong parameters. I have worked this around by grabbing the HTML page source and making it into a list of local server's URLs to images that work (I can open them in my browser). Saved the list to a different `.txt` file;
4. I selected the new `.txt` file at import and said to treat it as multiple tasks;
5. It seems to import ok, I can see thumbnails of my pictures, but when I select them for labeling - the picture sometimes shows for a split second and it disappears and I see an error message instead:
![image](https://github.com/HumanSignal/label-studio/assets/491704/53bda3ff-138f-47f8-9b2e-3ba70f8a9db0)

On the other side I see the local web server providing the files without errors:
```
Running web server on the port 9002
Serving HTTP on 0.0.0.0 port 9002 (http://0.0.0.0:9002/) ...
127.0.0.1 - - [13/Nov/2023 21:34:08] "GET /IMG-20221223-WA0001.jpeg HTTP/1.1" 200 -
127.0.0.1 - - [13/Nov/2023 21:34:47] "GET /IMG-20221203-WA0008.jpeg HTTP/1.1" 200 -
127.0.0.1 - - [13/Nov/2023 21:34:49] "GET /IMG-20221213-WA0006.jpeg HTTP/1.1" 200 -
127.0.0.1 - - [13/Nov/2023 21:34:50] "GET /IMG-20221126-WA0014.jpeg HTTP/1.1" 200 -
```

The label-studio executable also doens't print any errors in the console:
```
[2023-11-13 20:34:49,206] [django.server::log_message::161] [INFO] "GET /api/tasks/6?project=1 HTTP/1.1" 200 730
[2023-11-13 20:34:49,206] [django.server::log_message::161] [INFO] "GET /api/tasks/6?project=1 HTTP/1.1" 200 730
[2023-11-13 20:34:49,228] [django.server::log_message::161] [INFO] "GET /api/label_links?project=1&expand=label HTTP/1.1" 200 52
[2023-11-13 20:34:49,228] [django.server::log_message::161] [INFO] "GET /api/label_links?project=1&expand=label HTTP/1.1" 200 52
[2023-11-13 20:34:50,203] [django.server::log_message::161] [INFO] "GET /api/tasks/1?project=1 HTTP/1.1" 200 730
[2023-11-13 20:34:50,203] [django.server::log_message::161] [INFO] "GET /api/tasks/1?project=1 HTTP/1.1" 200 730
[2023-11-13 20:34:50,231] [django.server::log_message::161] [INFO] "GET /api/label_links?project=1&expand=label HTTP/1.1" 200 52
[2023-11-13 20:34:50,231] [django.server::log_message::161] [INFO] "GET /api/label_links?project=1&expand=label HTTP/1.1" 200 52
```

Maybe this here issue could be re-opened and renamed to a more general "Issues with using local images"? 

---

EDIT (an hour later)

I have somehow managed to get it working.
I have made sure my "local media root" path is one step above the place where I actually store photos, and then adding the local storage I added the longer absolute path and it worked!

I now first tried to import via a URL of an image, but that was one-at-a-time. I've found that I can now "upload" the image files directly from the path and that... worked?

I was able to label by (small) dataset of images! Woo!

**Comment by AlcobaAI** [comment]
Hi, 
This is how I got it to work, hope it helps :). The example is from what I was working on at the time.

1. Create a project and go to settings -> Cloud Storage: Add the location of the images and the output. Do not "Treat every bucket object as a source file" or sync. And check the connection.

![image](https://github.com/HumanSignal/label-studio/assets/53024349/0d558a77-8070-4b6d-88cc-7ffdb8e61466)


2. Label Interface:

```
<View>
  <Image name="image" value="$filename" zoom="true" zoomControl="true" maxWidth="100%"/>
  <Header size="10" value="$source_en"/>
  <Header size="10" value="$caption"/>

  <View>
  <Text name="q1" value=""/>
   <Header size="10" value=" Is the translation good? If so answer "Yes", if not write an alternate translation. />
    <TextArea name="answer1" toName="q1" rows="5" maxSubmissions="1"/>
  </View>
</View>
```

3. Here I had a separate CSV file with "filename" and 2 other columns "source_en" and "caption". In the filename add each separate path as "data/local-files/?d=/home/pathtoimages/image_name1.jpg". And import the CSV.

![image](https://github.com/HumanSignal/label-studio/assets/53024349/f65c3432-291f-4fa4-8b75-70ef40c89554)

**Comment by SCZwangxiao** [comment]
Here is the detailed explanation: https://labelstud.io/guide/storage.html#Local-storage

**Comment by Trevol** [comment]
I also searched for simple option to say Label Studio to take all images in specified directory...
But managed to import images from local storage in following way:
1. My Label Studio runs as docker container. So my local images directory mounted as volumes and necessary environment was specified. Snippet from docker-compose.yml
`  environment:            
      - LABEL_STUDIO_LOCAL_FILES_SERVING_ENABLED=true
      - LABEL_STUDIO_LOCAL_FILES_DOCUMENT_ROOT=/label-studio/data/images    
    volumes:      
      - "./label-studio-data:/label-studio/data"
      - **"./images:/label-studio/data/images"**`

2. Add Source Cloud Storage with type "Local files" and path /label-studio/data/images/pylons. Subdirectory pylons contains image files for annotation (object detection template). Now individual images become accessible via url {LABEL_STUDIO_HOST}/data/local-files?d=pylons/{IMAGE_FILE}  (in may case http://localhost:8080/data/local-files?d=pylons/1.jpeg)

3. Created (and then imported to Data Manager) single text file (pylons.txt in my case) with urls to local images:  
[pylons_sample.txt](https://github.com/user-attachments/files/17115994/pylons_sample.txt)
![Screenshot from 2024-09-24 17-10-25](https://github.com/user-attachments/assets/3fe693b0-4c29-4311-be6f-f024e7878bea)

**Comment by davidnrose** [comment]
Importing local files should not be this convoluted. Lots of time wasted

**Comment by jipijipi** [comment]
For OSX if anyone is as lost as I was : 

1) In Terminal _before_ launching Label Studio : 

- `export LABEL_STUDIO_LOCAL_FILES_SERVING_ENABLED=true`
- `export LOCAL_FILES_DOCUMENT_ROOT=/Users/me/parent_folder` (use your own path but not all the way to the final folder or it will fail later)

2) In Label Studio UI, in your project : 
- navigate to settings > cloud storage > local storage > add source storage and input the same path but include the child folder(s) this time : `/Users/me/parent_folder/image_folder`
- Click "add storage" and if it worked you'll know
![Screenshot 2024-11-05 at 23 00 06](https://github.com/user-attachments/assets/a8853019-ce42-44ec-893e-497be723826b)

3) In your JSON : 
- start with  `/data/local-files/?d=` followed by the relative path including the last folder of the path you declared in your source, example : 
` "data": {
            "image": "/data/local-files/?d=image_folder/image_sub_folder/image1.png"
        },`

TLDR;

If your absolute path looks like this : 
`/Users/me/parent_folder/image_folder/image_sub_folder/image1.png`
then :
`/Users/me/parent_folder/` This goes in environment variable (Terminal)
`/Users/me/parent_folder/image_folder` This goes in your storage source (UI)
`/data/local-files/?d=image_folder/image_sub_folder/image1.png` This goes in your JSON

**Comment by busedagidir** [comment]
> For OSX if anyone is as lost as I was :
> 
> 1. In Terminal _before_ launching Label Studio :
> 
> * `export LABEL_STUDIO_LOCAL_FILES_SERVING_ENABLED=true`
> * `export LOCAL_FILES_DOCUMENT_ROOT=/Users/me/parent_folder` (use your own path but not all the way to the final folder or it will fail later)
> 
> 2. In Label Studio UI, in your project :
> 
> * navigate to settings > cloud storage > local storage > add source storage and input the same path but include the child folder(s) this time : `/Users/me/parent_folder/image_folder`
> * Click "add storage" and if it worked you'll know
>   ![Screenshot 2024-11-05 at 23 00 06](https://private-user-images.githubusercontent.com/27646008/383337628-a8853019-ce42-44ec-893e-497be723826b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzE1MDMxNjksIm5iZiI6MTczMTUwMjg2OSwicGF0aCI6Ii8yNzY0NjAwOC8zODMzMzc2MjgtYTg4NTMwMTktY2U0Mi00NGVjLTg5M2UtNDk3YmU3MjM4MjZiLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMTMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTEzVDEzMDEwOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWExNTgwNzZmOTQxNzE2MzkzN2I5YjQwNGQ5N2MyNzI3NmExMTI1Y2VjNWVkMjliOTgwNzY3MWIzMTI3OTI1ZjImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.Mb17gqCuHwZimUggcC85S0nJDsHgPHYCA5estUfDPYE)
> 
> 3. In your JSON :
> 
> * start with  `/data/local-files/?d=` followed by the relative path including the last folder of the path you declared in your source, example :
>   ` "data": { "image": "/data/local-files/?d=image_folder/image_sub_folder/image1.png" },`
> 
> TLDR;
> 
> If your absolute path looks like this : `/Users/me/parent_folder/image_folder/image_sub_folder/image1.png` then : `/Users/me/parent_folder/` This goes in environment variable (Terminal) `/Users/me/parent_folder/image_folder` This goes in your storage source (UI) `/data/local-files/?d=image_folder/image_sub_folder/image1.png` This goes in your JSON

is this the only solution and really works?

**Comment by nina-link-u** [comment]
Didn't work for me.

**Comment by folkiesss** [comment]
I wasted my time half day to figure it out. so frustrated.

I setup with Docker compose. with this directory structure.
```
/home/username/label-studio-project/    <- (Where your compose.yaml is)
├── compose.yaml
└── local-files/            <- Your local data directory
    └── images/
        ├── import.json
        ├── file1.jpg
        ├── file2.jpg
        └── ... (additional files)
    └── other-dir/
        ├── import.json
        ├── file1.jpg
        ├── file2.jpg
        └── ... (additional files)
```

I need to add these to my compose file
```yaml
    environment:
      .....
      # add these 2 variable
      - LOCAL_FILES_SERVING_ENABLED=true
      - LABEL_STUDIO_LOCAL_FILES_DOCUMENT_ROOT=/data/local-files  # <- This will be you absolute path in what on UI said
    volumes:
      - label-studio-data:/label-studio/data:rw
      - ./local-files:/data/local-files   # <-- add your folder as bind mount
```

on UI you put

<img width="842" height="574" alt="Label Studio Import Dialog" src="https://github.com/user-attachments/assets/ca80ff42-1d9a-4c89-ae01-809c9f67e583" />

Then you can filter for import.json

> [!IMPORTANT]
> For file location you need to put `/data/local-files?d=<directoryName>/<filename>`.
> The `/data/local-files` prefix is **FIXED** it doesn't need to match the `LABEL_STUDIO_LOCAL_FILES_DOCUMENT_ROOT` environment variable, and regardless of your bind mount directory name.
> 
> For example, even if you rename your directory:
> ```
> /home/username/label-studio-project/
> ├── compose.yaml
> └── my-custom-folder/  <-- Any name works here
>     ├── images/
>     └── other-dir/
> ```
> You **must still use** `/data/local-files?d=...` in the file paths within Label Studio.

Now your files will be accessible at:

```
<your-label-studio-url>/data/local-files?d=<subdirectory>/<filename>
```

hope this help.
